### ISSUE TITLE
[Clang] `no_unique_address` is reserved word in pre-C++20 dialects (non-conforming)

### ISSUE THREAD

**Issue by JoeLoser** [llvm/llvm-project#61196]
As we can see on [this Godbolt link](https://godbolt.org/z/5W4EoEM4E), even when compiling in earlier dialects than C++20 (such as C++17), we get a compiler error since `no_unique_address` is effectively a reserved word even in pre-C++20 dialects.

Some context: Clang supports `[[no_unique_address]` in all dialects for C++11 and later (I don't know why it doesn't support it in C++03 mode, even as `__attribute__((no_unique_address))`).

Failing to support a double-underscored `no_unique_address` means that `no_unique_address` encroaches into the user's namespace in C++11 through C++17 dialects which is not good.

One resolution would be if Clang could start supporting a __dundered form of `no_unique_address` that would be safely usable by the implementor, e.g. `__attribute__((__no_unique_address__))`.

**Comment by shafik** [comment]
As per documentation: https://clang.llvm.org/docs/AttributeReference.html#no-unique-address this is supported in C++11 forward. I am looking at the PR: https://reviews.llvm.org/D63451 that implemented it and I don't see a rationale for supporting it as a extension in C++11.

CC @zygoloid who implemented and can perhaps talk to why it is supported pre C++20 and attributes code owner @erichkeane

**Comment by zygoloid** [comment]
I'm not sure I understand the question. It's clearly useful to support this in older versions, and it's conforming to do so. Indeed, the design of C++ attributes is intended precisely to permit compilers supporting attributes in older versions.

There's definitely a libc++ bug here -- it's not conforming for it to use `__has_cpp_attribute(msvc::no_unique_address)` like this because neither `msvc` nor `no_unique_address` are on the list of identifiers that a program shall not define before including a library header.

Now, Clang does support `__no_unique_address__` as a synonym for `no_unique_address`, both in the attribute and in the `__has_cpp_attribute` check, but using that doesn't help, because this is a check for an MSVC feature, so what matters is what msvc supports.

Presumably the proper fix for libc++ would be to guard this check with something like

```
#if !defined(msvc) && !defined(no_unique_address)
```

... but I suppose there's still the possibility that the user will `#define no_unique_address` or `msvc` between including the config header and including some other library header, so maybe that doesn't work?

In any case, I don't see any Clang issue here. Retargeting as a libc++ bug.

**Comment by shafik** [comment]
@zygoloid maybe I am reading it wrong but it seems like [macro.names](https://eel.is/c++draft/macro.names#2) would reserve the attributes names and by supporting it as a extension we are encroaching on the user namespace early then folks would expect.

**Comment by philnik777** [comment]
> @zygoloid maybe I am reading it wrong but it seems like [macro.names](https://eel.is/c++draft/macro.names#2) would reserve the _attribute-tokens_ and by supporting it as a extension we are encroaching on the user namespace early then folks would expect.

The _attribute-token_ s are the standard attributes, i.e. you can't `#define no_unique_address` in C++20-and-later, but you are allowed to do that in C++17. This essentially just means that the library is allowed to use the non-_Uglified versions of the attributes when they are in the standard. So you are free to do fun things like [this](https://godbolt.org/z/sv44zP4vr), and as you see the compiler is happy to do that.

I think we should just remove the `[[msvc::no_unique_address]]` for now, since it's not supported by clang or GCC currently anyways.

**Comment by shafik** [comment]
I chatted w/ @zygoloid offline and we are saying the same thing, I was just missing some context. So I agree with @zygoloid original comment.

**Comment by frederick-vs-ja** [comment]
> There's definitely a libc++ bug here -- it's not conforming for it to use `__has_cpp_attribute(msvc::no_unique_address)` like this because neither `msvc` nor `no_unique_address` are on the list of identifiers that a program shall not define before including a library header.

MSVC STL resolved this by preprocesor extensions (microsoft/STL#2649). Is the way suitable for libc++?

**Comment by philnik777** [comment]
> > There's definitely a libc++ bug here -- it's not conforming for it to use `__has_cpp_attribute(msvc::no_unique_address)` like this because neither `msvc` nor `no_unique_address` are on the list of identifiers that a program shall not define before including a library header.
> 
> MSVC STL resolved this by preprocessor extensions ([microsoft/STL#2649](https://github.com/microsoft/STL/pull/2649)). Is the way suitable for libc++?

We don't support the Microsoft compiler anyways, so we can just ask the clang folks to add an `__msvc__` spelling if `[[msvc::no_unique_address]]` ever gets implemented.

**Comment by zygoloid** [comment]
> > There's definitely a libc++ bug here -- it's not conforming for it to use `__has_cpp_attribute(msvc::no_unique_address)` like this because neither `msvc` nor `no_unique_address` are on the list of identifiers that a program shall not define before including a library header.
> 
> MSVC STL resolved this by preprocessor extensions ([microsoft/STL#2649](https://github.com/microsoft/STL/pull/2649)). Is the way suitable for libc++?

MSVC's approach is still broken: https://godbolt.org/z/cTYor9Ta6

It's also unsuitable for `[[msvc::no_unique_address]]` (unless the fallback path gets the same class layout a different way), because falling back on not using the attribute could result in an ABI difference between different TUs in the same program.

**Comment by zygoloid** [comment]
Reopening: the original issue reported here hasn't been fixed yet.

**Comment by frederick-vs-ja** [comment]
I think I've roughly completed MSVC's approach (microsoft/STL#3760).

If we apply that approach to libc++, there'll 35 files to be modified now ([search result](https://github.com/search?q=repo%3Allvm%2Fllvm-project+path%3Alibcxx%2Finclude+_LIBCPP_NO_UNIQUE_ADDRESS&type=code))...

**Comment by philnik777** [comment]
> I think I've roughly completed MSVC's approach ([microsoft/STL#3760](https://github.com/microsoft/STL/pull/3760)).
> 
> If we apply that approach to libc++, there'll 35 files to be modified now ([search result](https://github.com/search?q=repo%3Allvm%2Fllvm-project+path%3Alibcxx%2Finclude+_LIBCPP_NO_UNIQUE_ADDRESS&type=code))...

The attributes themselves are already _Ugly. The only thing that hasn't been fixed is when checking for attribute availability.

**Comment by frederick-vs-ja** [comment]
> The attributes themselves are already _Ugly. The only thing that hasn't been fixed is when checking for attribute availability.

However, when the non-_Uglified attribute token `msvc::no_unique_address` is actually used (although currently it is never), `#define msvc sth` can still be problematic.

`no_unique_address` is perhaps OK itself since the attributes are unlikely to be used in pre-C++20 modes.

**Comment by philnik777** [comment]
> > The attributes themselves are already _Ugly. The only thing that hasn't been fixed is when checking for attribute availability.
> 
> However, when the non-_Uglified attribute token `msvc::no_unique_address` is actually used (although currently it is never), `#define msvc sth` can still be problematic.
> 
> `no_unique_address` is perhaps OK itself since the attributes are unlikely to be used in pre-C++20 modes.

As I said before, IMO Clang should just accept some _Ugly version of the msvc attributes. That's how it's done for gnu and clang attributes, and makes this all trivial to fix.

**Comment by cor3ntin** [comment]
@philnik777 we do not support the attribute at all (yet), so it is difficile to preemptively do that. and actually, we can't do that for any attributes we don't support as it is an open set

**Comment by philnik777** [comment]
> @philnik777 we do not support the attribute at all (yet), so it is difficile to preemptively do that. and actually, we can't do that for any attributes we don't support as it is an open set
> 

Yes, I'm aware of that. I just meant that for any attribute that clang accepts it should also accept an _Ugly alias, since that doesn't look like a ton of effort to me for the compiler but makes the life of standard library maintainers so much easier.

**Comment by zygoloid** [comment]
This is not a clang issue, the problem is in libc++. Clang changes won't help.

**Comment by philnik777** [comment]
> This is not a clang issue, the problem is in libc++. Clang changes won't help.

Yes, this is a libc++ issue. But I don't want to push and pop macros all over the place once clang supports `[[msvc::no_unique_address]]`. There should just be `[[__msvc__::__no_unique_address__]]` just like there are `[[_Clang::__whatever__]]` and `[[__gnu__::__whatever__]]` so libraries have it easier.

**Comment by zygoloid** [comment]
Clang already supports that for all its attributes. The issue is that MSVC does not, so if we want to support MSVC, then that doesn't help us. If we don't want to support MSVC then we should just remove all the `msvc::no_unique_address` code from libc++ to fix this bug. Then we can implement something clang-specific once #65675 is merged, but that's not part of this bug.

**Comment by cor3ntin** [comment]
Agreed. TBH, @frederick-vs-ja  solution of using pragma push seems like the best solution here.
